go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -433,3 +447,191 @@ func (call *builtinWeightString) compile(c *compiler) (ctype, error) {
 		return ctype{Type: sqltypes.VarBinary, Flag: flagNullable | flagNull, Col: collationBinary}, nil
 	}
 }
+
+func (call builtinLeftRight) eval(env *ExpressionEnv) (eval, error) {
+	str, l, err := call.arg2(env)
+	if err != nil {
+		return nil, err
+	}
+	if str == nil || l == nil {
+		return nil, nil
+	}
+
+	text, ok := str.(*evalBytes)
+	if !ok {
+		text, err = evalToVarchar(str, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	length := evalToInt64(l).i
+	if length <= 0 {
+		return newEvalText(nil, text.col), nil
+	}
+
+	cs := text.col.Collation.Get().Charset()
+	strLen := charset.Length(cs, text.bytes)
+
+	if strLen <= int(length) {
+		return newEvalText(text.bytes, text.col), nil
+	}
+
+	var res []byte
+	if call.left {
+		res = charset.Slice(cs, text.bytes, 0, int(length))
+	} else {
+		res = charset.Slice(cs, text.bytes, strLen-int(length), strLen)
+	}
+	return newEvalText(res, text.col), nil
+}
+
+func (call builtinLeftRight) typeof(env *ExpressionEnv, fields []*querypb.Field) (sqltypes.Type, typeFlag) {
+	_, f1 := call.Arguments[0].typeof(env, fields)
+	return sqltypes.VarChar, f1
+}
+
+func (call builtinLeftRight) compile(c *compiler) (ctype, error) {
+	str, err := call.Arguments[0].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	l, err := call.Arguments[1].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	skip := c.compileNullCheck2(str, l)
+
+	col := defaultCoercionCollation(c.cfg.Collation)
+	switch {
+	case str.isTextual():
+		col = str.Col
+	default:
+		c.asm.Convert_xc(2, sqltypes.VarChar, col.Collation, 0, false)
+	}
+	_ = c.compileToInt64(l, 1)
+
+	if call.left {
+		c.asm.Fn_LEFT(col)
+	} else {
+		c.asm.Fn_RIGHT(col)
+	}
+	c.asm.jumpDestination(skip)
+	return ctype{Type: sqltypes.VarChar, Col: col, Flag: flagNullable}, nil
+}
+
+func (call builtinPad) eval(env *ExpressionEnv) (eval, error) {
+	str, l, p, err := call.arg3(env)
+	if err != nil {
+		return nil, err
+	}
+
+	if str == nil || l == nil || p == nil {
+		return nil, nil
+	}
+
+	text, ok := str.(*evalBytes)
+	if !ok {
+		text, err = evalToVarchar(str, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	length := evalToInt64(l).i
+	if length < 0 {
+		return nil, nil
+	}
+
+	pad, ok := p.(*evalBytes)
+	if !ok {
+		pad, err = evalToVarchar(p, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !checkMaxLength(int64(len(pad.bytes)), length) {
```

Comment:
Nit, but IMO a better function name would be `wouldBeBeyondMaxLen()` with the bool value returned being flipped (obviously). Or a smaller change, name only, to `isValidMaxLength()`

Reply:
I already have a follow up for this PR scheduled with more functions added, so going to change it there to avoid the CI churn.

Reply:
See https://github.com/vitessio/vitess/pull/13024